### PR TITLE
Fix KVO observation in @StateRealmObject

### PR DESCRIPTION
**Pull request by leemaguire**
# Problem
Currently if you try to add an unmanaged object that is declared with `@StateRealmObject` to a Realm or collection you will get an exception stating that objects with observers cannot be added to the Realm. 

# Solution
This PR addresses that issue by adding a check inside `RLMAccessorContext::createObject` which sees if the object being added to the Realm has any existing `Subscriptions` in `SwiftUIKVO`, if there are then we remove the observers for that object and once the object is successfully managed by the Realm we add back a set of new observers for the properties on that object. The new observers will continue to use the same `SwiftUIKVO` object for receiving the notifications.

I don't see a need to use KVO context pointers here but maybe I'm wrong.

This PR also sneaks in the doc change for https://github.com/realm/realm-cocoa/issues/7390

### CODE REVIEW

File: Realm/RLMAccessor.mm [hunk 1]
```diff
@@ -1034,6 +1034,7 @@ static auto toOptional(__unsafe_unretained id const value) {
 
     RLMObjectBase *objBase = RLMDynamicCast<RLMObjectBase>(value);
     realm::Obj obj, *outObj = nullptr;
+    bool requiresSwiftUIObservers;
```
Comment (tgoyne) [inline Realm/RLMAccessor.mm:1037]: This variable needs to be initialized to a value.

File: Realm/RLMObservation.hpp [hunk 2]
```diff
@@ -188,3 +188,12 @@ void RLMDidChange(std::vector<realm::BindingContext::ObserverState> const& obser
 realm::KeyPathArray RLMKeyPathArrayFromStringArray(RLMRealm *realm,
                                                    RLMClassInfo *info,
                                                    NSArray<NSString *> *keyPath);
+
+// Used for checking if `Object`s declared with `@StateRealmObject` need to have
+// their accessors temporarily removed and added back so that the `Object` can be
+// managed be the Realm.
+[[clang::objc_runtime_visible]]
+@interface _SwiftUIKVO : NSObject
```
Comment (tgoyne) [inline Realm/RLMObservation.hpp:196]: This is not a legal name for a type in C. All names starting with an underscore and then followed by a capital letter are reserved for the standard library. It should be RLMSwiftUIKVO.

File: Realm/RLMObservation.hpp [hunk 3]
```diff
@@ -188,3 +188,12 @@ void RLMDidChange(std::vector<realm::BindingContext::ObserverState> const& obser
...
++ (BOOL)removeObserversWithObject:(NSObject *)object;
++ (void)addObserversWithObject:(NSObject *)object;
```
Comment (tgoyne) [inline Realm/RLMObservation.hpp:198]: "FromObject" and "ToObject" rather than "With"

File: RealmSwift/SwiftUI.swift [hunk 4]
```diff
@@ -67,41 +67,64 @@ private func createBinding<T: ThreadConfined, V>(_ value: T,
     })
 }
 
-// MARK: SwiftUIKVO
+// MARK: _SwiftUIKVO
 
+// swiftlint:disable type_name
 @available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.0, *)
-internal final class SwiftUIKVO: NSObject {
+@objc(_SwiftUIKVO) public class _SwiftUIKVO: NSObject {
```
Comment (tgoyne) [inline RealmSwift/SwiftUI.swift:74]: This should still be internal and named SwiftUIKVO.

**Review by jsflax — APPROVED** [review]

**Review by tgoyne — APPROVED** [review]

**Pull request merged**